Add HTA-side implementation of sync-interrupted scenario (HT-508). Also replace deprecated AsyncTask.

wmergenthal opened the pull request:
Formerly HT sent HTA "sync_success" regardless of whether 'mergeCompleted' was True. Now the False case leads to HTA receiving sync status "sync_interrupted". HTA parses the HttpRequest from HT to extract the status, then informs the user. One new string is added to HTA for the interrupt-or-cancelled case.
Also, deprecated class 'AsyncTask' is replaced with 'Executors' and 'Handlers'.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13)
<!-- Reviewable:end -->

<!-- devin-review-badge-begin -->

---

<a href="https://app.devin.ai/review/sillsdev/hearthisandroid/pull/13" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    <img src="https://static.devin.ai/assets/gh-open-in-devin-review-light.svg?v=1" alt="Open with Devin">
  </picture>
</a>
<!-- devin-review-badge-end -->

wmergenthal committed:
Handle both 'sync_success' and 'sync_interrupted' statuses from HT instead of hardcoding "success". Replace deprecated AsyncTask.

Added handling for the HttpRequest from HT which now can have either "sync_success" or "sync_interrupted". Formerly the user would see a success message in either case. Now, however, when a sync is cancelled the user will see a message indicating that sync did not complete.
Replaced AsyncTask (which is deprecated) with Executors and Handlers.
This still contains debug output from the investigation activities. A future commit (very soon, hopefully) will remove this in PR preparation.

wmergenthal committed:
HT-508: finish HTA-side implementation of sync-interrupted scenario. Also replace deprecated AsyncTask code.

Formerly HT sent HTA "sync_success" regardless of whether 'mergeCompleted' was True. Now the False case leads to HTA receiving "sync_interrupted" and HTA informs the user that sync did not complete successfully.
Also, deprecated class 'AsyncTask' is replaced with 'Executors' and 'Handlers'.

tombogle reviewed (changes requested):
> @tombogle reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13)* status: all files reviewed, 4 unresolved discussions (waiting on @wmergenthal)

___
*[`app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java` line 43 at r1](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-OZ9xdxf4M7xxprgIyG1:-OZ9xdxf4M7xxprgIyG2:bi9jpyh) ([raw file](https://github.com/sillsdev/HearThisAndroid/blob/fd35b080e346d41d75e8425714fd7401daa77b5c/app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java#L43)):*
> ```Java
>         //
>         // NOTE: like several things in HearThisAndroid, HttpRequest is deprecated. It will be
>         // replaced with something more appropriate, hopefully soon.
> ```

Options:
-   **Use `HttpURLConnection` (built-in JDK/Android)**  
    → Standard replacement for most simple request/response work.
    
    -   Pros: No external dependency.
        
    -   Cons: Verbose API; no built-in high-level features.
        
-   **Use `OkHttp` (square.github.io/okhttp)**  
    → De facto standard for Android networking.
    
    -   Pros: Clean API, better performance, supports HTTP/2, widely adopted.
        
    -   Cons: Adds dependency, but well worth it in most apps.
___
*[`app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java` line 49 at r1](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-OZ9y_oM3S9HKaknu3Dr:-OZ9y_oM3S9HKaknu3Ds:b-mfzhe9) ([raw file](https://github.com/sillsdev/HearThisAndroid/blob/fd35b080e346d41d75e8425714fd7401daa77b5c/app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java#L49)):*
> ```Java
>         String s2 = s1.substring(s1.indexOf(' ') + 1, s1.lastIndexOf(' '));
>         String status = s2.substring(s2.indexOf('=') + 1);
> ```

This logic could be done in a less cryptic way:
String s1 = request.getRequestLine().getUri(); // avoid toString() if possible
URI uri = new URI(s1);
String query = uri.getQuery(); // e.g., "status=123&foo=bar"

String status = null;
for (String param : query.split("&")) {
    String[] pair = param.split("=");
    if (pair.length == 2 && pair[0].equals("status")) {
        status = pair[1];
        break;
    }
}

Alternatively, it could be done very simply with a regular expression. But using URI is probably safer and makes the intent clear.
___
*[`app/src/main/java/org/sil/hearthis/SyncActivity.java` line 166 at r1](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-OZ9zRDT9rKYFhvXaqdW:-OZ9zRDT9rKYFhvXaqdX:b-4if43u) ([raw file](https://github.com/sillsdev/HearThisAndroid/blob/fd35b080e346d41d75e8425714fd7401daa77b5c/app/src/main/java/org/sil/hearthis/SyncActivity.java#L166)):*
> ```Java
>                                                               socket.send(packet);
>                                                           } catch (UnknownHostException e) {
>                                                               e.printStackTrace();
> ```

In the case of an exception, is there anything else we need to do to give feedback to the user and/or prevent the HTA app from getting stuck in a bad state?
___
*[`app/src/main/java/org/sil/hearthis/SyncActivity.java` line 280 at r1](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-OZ9zvsjDDDDWFhXvPeW:-OZ9zvsjDDDDWFhXvPeX:bc90ik4) ([raw file](https://github.com/sillsdev/HearThisAndroid/blob/fd35b080e346d41d75e8425714fd7401daa77b5c/app/src/main/java/org/sil/hearthis/SyncActivity.java#L280)):*
> ```Java
>             setProgress(getString(R.string.sync_interrupted));
>         } else {
>             // Should never happen. Not sure what to do here, if anything...
> ```

For the sake of future proofing, we should probably handle any unexpected message by letting the use know that they are probably using a version of HTA that is not fully compatible with HT desktop. Then *probably* we should treat it similar to the "interrupted" case, since we probably can't safely assume it was completed successfully.


<!-- Sent from Reviewable.io -->

wmergenthal committed:
Sync final status extracted from HT's notification with greater clarity

Instead of raw string objects, use a URI object for status extraction from the HttpRequest from HT. Resulting logic is longer but clearer.

wmergenthal committed:
Handle case when HTA does not understand sync final status from HT

Treat similarly to when sync is interrupted. Show the user a non-success message and suggest that they verify that HT/HTA versions are close enough that HTA knows about all possible HT sync statuses.

wmergenthal commented:
Re comment 1: thanks for suggesting some options. OkHttp certainly looks compelling.
Re comment 2: addressed in 3727d83.
Re comment 3: currently working on this one.
Re comment 4: addressed in 399e532.

tombogle commented on app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java:
```diff
@@ -42,10 +43,26 @@ public void handle(HttpRequest request, HttpResponse response, HttpContext httpC
+
+        if (status == null) {
+            // Something went wrong. Make sure the user sees a non-success message.
+            status = "sync_interrupted";
```
I wonder if we should display a different message in the case where we get an unexpected status message. I assume the processing should be the same (since we can't possibly know how else to interpret the unexpected message), but if this ever happens, we could give the user some indication that we got an unexpected message and suggest they check for an update. We could even include in the query (from HT) an indication of the minimum version of HTA known to handle the particular status message. Then if we ever get an unexpected status and it includes a version number as well, we could tell the user the minimum version of HTA that is compatible with the version of HT they are using. The HTA code would then look something like this:
```
String status = null;
String minHtaVersion = null;
try {
    String s1 = request.getRequestLine().getUri();
    URI uri = new URI(s1);
    String query = uri.getQuery();
    if (query != null) {
        for (String param : query.split("&")) {
            String[] pair = param.split("=", 2); // limit=2 in case value contains '='
            if (pair.length == 2) {
                if (pair[0].equals("message")) {
                    status = pair[1];
                } else if (pair[0].equals("minHtaVersion")) {
                    minHtaVersion = pair[1];
                }
            }
        }
    }
} catch (Exception e) {
    e.printStackTrace();
}

if (status == null || isUnexpectedStatus(status)) {
    if (minHtaVersion != null) {
        showError("Unexpected sync status received. This version of HearThis for Android appears to be incompatible with your version of HearThis. Please update HearThis for Android to at least version " + minHtaVersion + ".");
    } else {
        showError("Unexpected sync status received. Please check for updates.");
    }
    status = "sync_interrupted"; // Or possibly a distinct status
}
```

wmergenthal replied:
Same comment as for the flagged 'minHtaVersion' item above -- I want to leave it as is for now because the 'minHtaVersion' bit of mechanism is only partially implemented.
<!-- Reviewable comment -Omgb1YKEgxS6SgvNe1q:bpxoocb -->
<!-- Sent from Reviewable.io -->

wmergenthal committed:
Safety checkin: clearer status extract, add watchdog and debug output

Nominal case works. Still need to test watchdog when HT fails to complete its side of sync. Debug output still present (which helps greatly understanding the mechanism).

wmergenthal committed:
Safety commit: watchdog works, debug info shows sync activities

No substantive changes. Just captures debug stmts producing the output captured in 20251001_02_HTA_sync_watchdog_all_ok.log, where watchdog is seen to work correctly for both success and fail scenarios.

wmergenthal committed:
watchdog: start and stop in better ways

Watchdog was being started too early. If the Android user was not quick to scan the QR code after getting into that mode, the watchdog could have timed out before the sync even started. Now we don't start it until Android has sent its UDP packet to PC to initiate a sync.
Also, multiple watchdog shutdown() calls are now consolidated into a single one.
Add comments to the new Watchdog class.
Remove some dead code and improve some debug output stmts.

wmergenthal committed:
Ungray the Continue button upon watchdog timeout

Without the Continue button the user can't get back to the home screen unless a subsequent sync retry is successful.
To make the Continue button usable, add call to AcceptNotificationHandler like all the other spots have.

tombogle commented:
*[`app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java` line 85 at r3](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-OjbbP3R3XnKp3pYUlE_:-OjbbP3R3XnKp3pYUlEa:bh6jbun) ([raw file](https://github.com/sillsdev/HearThisAndroid/blob/fd787610ed31bdb609c29d621bf4426f65ff070c/app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java#L85)):*
> ```Java
>         if (status == null) {
>             // We got something but it wasn't "status". Make sure the user sees an error message.
>             status = "sync_error";
> ```

From an error handling/reporting standpoint, if we did get a query, it might be a good idea to display and/or log it in a way that it could be reported to us so we could try to figure out what happened.


<!-- Sent from Reviewable.io -->

tombogle reviewed (commented):
> @tombogle reviewed 2 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13)* status: 2 of 7 files reviewed, 6 unresolved discussions (waiting on @wmergenthal).



<!-- Sent from Reviewable.io -->

wmergenthal committed:
Change sync status 'interrupted' to 'canceled'

This is a coordinating change with the HT side. Last fall we had decided to change a few wordings, one of which is to change "sync_interrupted" to "sync_canceled". This comm does that as well as commenting out several debug output stmts. Debug output will be removed altogether before the next push, but this is an intermediate snapshot commit.

wmergenthal committed:
Prep for PR update: remove or disable all debug output

There are a handful of debug output things I want to capture in a way that enables easy and reliable transmission to our tech support. These are commented out along with a notation to "implement for tech support".
(1) Timeout reduction, (2) minHtaVersion and (3) exception handling fixes are being postponed until after the Code-a-thon. Some of these might get done *during* the Code-a-thon, but even if they don't it will be preferable to develop these 3 things on the updated code base.
The intent of this commit is to complete all needed changes requested for Pull Request #13.

andrew-polk commented:
I hope this is helpful and doesn't just add more noise, but I now trust devin review more than a human to do reviews.
We don't have things hooked up to make it automatic to report easily, but it found these. If you would like more detail on any of them, let me know.

https://app.devin.ai/review/sillsdev/HearThisAndroid/pull/13#issuecomment-3910931106

<img width="597" height="662" alt="image" src="https://github.com/user-attachments/assets/3d7bd758-9cc8-4365-af48-adf89b9baaa6" />

wmergenthal committed:
Fix 10 issues flagged by devin-bot

Watchdog.java - start timer as last action in constructor; make shutdown() synchronized like pet() is; remove unused import
SyncActivity.java - fix thread violation by moving ipView call up into calling thread; add socket close; add executor shutdown; add null checks to watchdog calls
AcceptNotificationHandler.java - add response body for minHtaVersion which is still only partially implemented

wmergenthal committed:
Remove unnecessary setProgress() call

Devin pointed out that this gets done as part of listener.onNotification().

wmergenthal commented on app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java:
```diff
@@ -32,9 +37,54 @@ public void handle(HttpRequest request, HttpResponse response, HttpContext httpC
+                    String[] pair = param.split("=", 2);  // limit=2 in case value contains '='
+                    if (pair.length == 2) {
+                        if (pair[0].equals("status")) {
+                            status = pair[1];
+                        } else if (pair[0].equals("minHtaVersion")) {
+                            minHtaVersion = pair[1];
+                            return;
```
I added an HTTP response as devin suggested. Seems to work fine.
<!-- Reviewable comment -OmgbTl-9A4CoIkqAccY:b-4woifw -->
<!-- Sent from Reviewable.io -->

wmergenthal commented on app/src/main/java/org/sil/hearthis/SyncActivity.java:
```diff
@@ -248,7 +291,33 @@ public boolean onOptionsItemSelected(MenuItem item) {
+        // is unable to complete a sync operation. Getting here means we got a notification
+        // from the PC. It should contain the final sync status, but even if it doesn't, the
+        // sync operation *is* complete and the watchdog should be turned off.
+        watchdog.shutdown();
```
Added null check.
<!-- Reviewable comment -OmgcE6N64ofaLw8IR20:b2m7sqi -->
<!-- Sent from Reviewable.io -->

wmergenthal commented on app/src/main/java/org/sil/hearthis/SyncActivity.java:
```diff
@@ -270,6 +339,8 @@ public void run() {
 
     @Override
     public void receivingFile(final String name) {
+        watchdog.pet();
```
Added null check.
<!-- Reviewable comment -OmgcZNuEEPNpohV4tIh:b2m7sqi -->
<!-- Sent from Reviewable.io -->

wmergenthal commented on app/src/main/java/org/sil/hearthis/Watchdog.java:
```diff
@@ -0,0 +1,44 @@
+    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
+    private ScheduledFuture<?> watchdogTask;
+    private final Runnable onTimeout;
+    private final long timeout;
+    private final TimeUnit unit;
+
+    public Watchdog(long timeout, TimeUnit unit, Runnable onTimeout) {
+        this.timeout = timeout;
+        this.unit = unit;
+        this.onTimeout = onTimeout;
+    }
```
Initially I started the countdown at the first transmission, but I can see the case for starting it earlier. Added a pet() call as the last thing in the watchdog constructor.
<!-- Reviewable comment -OmgcfKW13TzhH5E6fPF:b-4gw6et -->
<!-- Sent from Reviewable.io -->

wmergenthal reviewed (commented):
I addressed all items except the ones involving 'minHtaVersion'. That should be left as is for now since it is a new bit of the mechanism that is only partially implemented, to be finished asap.

> @wmergenthal made 11 comments.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-:-Omgabd67n5yqt5O4u9g:bbcjuq9)* status: 0 of 7 files reviewed, 10 unresolved discussions (waiting on tombogle).

___
*[`app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java` line 43 at r1](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-OZ9xdxf4M7xxprgIyG1:-Ol44MEW-os0jlo-kTD-:b7jk12s) ([raw file](https://github.com/sillsdev/HearThisAndroid/blob/fd35b080e346d41d75e8425714fd7401daa77b5c/app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java#L43)):*
<details><summary><i>Previously, tombogle (Tom Bogle) wrote…</i></summary><blockquote>

Options:
-   **Use `HttpURLConnection` (built-in JDK/Android)**  
    → Standard replacement for most simple request/response work.
    
    -   Pros: No external dependency.
        
    -   Cons: Verbose API; no built-in high-level features.
        
-   **Use `OkHttp` (square.github.io/okhttp)**  
    → De facto standard for Android networking.
    
    -   Pros: Clean API, better performance, supports HTTP/2, widely adopted.
        
    -   Cons: Adds dependency, but well worth it in most apps.
</blockquote></details>

I plan to leave this alone for now, and see if gets done during the Code-a-thon.
___
*[`app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java` line 49 at r1](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-OZ9y_oM3S9HKaknu3Dr:-OmgbhivCy6Uu7YG6TfL:bdn7doe) ([raw file](https://github.com/sillsdev/HearThisAndroid/blob/fd35b080e346d41d75e8425714fd7401daa77b5c/app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java#L49)):*
<details><summary><i>Previously, tombogle (Tom Bogle) wrote…</i></summary><blockquote>

This logic could be done in a less cryptic way:
String s1 = request.getRequestLine().getUri(); // avoid toString() if possible
URI uri = new URI(s1);
String query = uri.getQuery(); // e.g., "status=123&foo=bar"

String status = null;
for (String param : query.split("&")) {
    String[] pair = param.split("=");
    if (pair.length == 2 && pair[0].equals("status")) {
        status = pair[1];
        break;
    }
}

Alternatively, it could be done very simply with a regular expression. But using URI is probably safer and makes the intent clear.
</blockquote></details>

This has been implemented (with a slight modification).
___
*[`app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java` line 85 at r3](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-OjbbP3R3XnKp3pYUlE_:-OlchONX-ytUj7X_YiA6:b-ja12r) ([raw file](https://github.com/sillsdev/HearThisAndroid/blob/fd787610ed31bdb609c29d621bf4426f65ff070c/app/src/main/java/org/sil/hearthis/AcceptNotificationHandler.java#L85)):*
<details><summary><i>Previously, tombogle (Tom Bogle) wrote…</i></summary><blockquote>

From an error handling/reporting standpoint, if we did get a query, it might be a good idea to display and/or log it in a way that it could be reported to us so we could try to figure out what happened.
</blockquote></details>

Am postponing this until after the Code-a-thon. Or perhaps it will get done *during* that event, as a bonus, if things go well.
___
*[`app/src/main/java/org/sil/hearthis/SyncActivity.java` line 166 at r1](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-OZ9zRDT9rKYFhvXaqdW:-OmgbrD93DJT0HTqosWR:b-wro1u7) ([raw file](https://github.com/sillsdev/HearThisAndroid/blob/fd35b080e346d41d75e8425714fd7401daa77b5c/app/src/main/java/org/sil/hearthis/SyncActivity.java#L166)):*
<details><summary><i>Previously, tombogle (Tom Bogle) wrote…</i></summary><blockquote>

In the case of an exception, is there anything else we need to do to give feedback to the user and/or prevent the HTA app from getting stuck in a bad state?
</blockquote></details>

This will be an item for study after the Code-a-Thon when dropped connection error scenarios are considered.
___
*[`app/src/main/java/org/sil/hearthis/SyncActivity.java` line 280 at r1](https://reviewable.io/reviews/sillsdev/HearThisAndroid/13#-OZ9zvsjDDDDWFhXvPeW:-OmgcJVY2zbIdBnaDi29:bgmwtvt) ([raw file](https://github.com/sillsdev/HearThisAndroid/blob/fd35b080e346d41d75e8425714fd7401daa77b5c/app/src/main/java/org/sil/hearthis/SyncActivity.java#L280)):*
<details><summary><i>Previously, tombogle (Tom Bogle) wrote…</i></summary><blockquote>

For the sake of future proofing, we should probably handle any unexpected message by letting the use know that they are probably using a version of HTA that is not fully compatible with HT desktop. Then *probably* we should treat it similar to the "interrupted" case, since we probably can't safely assume it was completed successfully.
</blockquote></details>

Will be handled after the Code-a-Thon as part of the 'minHtaVersion' mechanism.


<!-- Sent from Reviewable.io -->
